pkg/controller/stack/stack_controller.go

```diff
@@ -285,6 +317,71 @@ func add(mgr manager.Manager, r *ReconcileStack) error {
 	return nil
 }
 
+// isRequirementSatisfied checks the given readiness requirement against the given stack, and
+// returns nil if the requirement is satisfied, and an error otherwise. The requirement can be nil
+// itself, in which case the prerequisite is only that the stack succeeded onm its last run.
+func isRequirementSatisfied(req *shared.RequirementSpec, stack pulumiv1.Stack) error {
+	if stack.Status.LastUpdate == nil {
+		return errRequirementNotRun
+	}
+	if stack.Status.LastUpdate.State != shared.SucceededStackStateMessage {
+		return errRequirementFailed
+	}
+	if req != nil && req.SucceededWithinDuration != nil {
+		lastRun := stack.Status.LastUpdate.LastResyncTime
+		if lastRun.IsZero() || time.Since(lastRun.Time) > req.SucceededWithinDuration.Duration {
+			return errRequirementOutOfDate
+		}
+	}
+	return nil
+}
+
+// ReconcileRequestedPredicate filters (returns true) for resources that are updated with a new or
+// amended annotation value at `ReconcileRequestAnnotation`.
+//
+// Reconciliation request protocol:
+//
+// This gives a means of prompting the controller to reconsider a resource that otherwise might not
+// be queued, e.g., because it has already reached a success state. This is useful for command-line
+// tooling (e.g., you can trigger updates with `kubectl annotate`), and is the mechanism used to
+// requeue prerequisites that are not up to date.
+//
+// The protocol works like this:
+
+//   - when you want the object to be considered for reconciliation, annotate it with the key
+//     `shared.ReconcileRequestAnnotation` and any likely-to-be-unique value. This causes the object
+//     to be queued for consideration by the controller;
+//   - the controller shall save the value of the annotation to `.status.observedReconcileRequest`
+//     whenever it processes a resource without returning an error. This is so you can check the
+//     status to see whether the annotation has been seen (similar to `.status.observedGeneration`).
+//
+// This protocol is the same mechanism used by many Flux controllers, as explained at
+// https://pkg.go.dev/github.com/fluxcd/pkg/runtime/predicates#ReconcileRequestedPredicate and
+// related documentation.
+type ReconcileRequestedPredicate struct {
+	predicate.Funcs
+}
+
+func getReconcileRequestAnnotation(obj client.Object) (string, bool) {
+	r, ok := obj.GetAnnotations()[shared.ReconcileRequestAnnotation]
+	return r, ok
```

Comment:
Do we need to destructure this return? Can we just return `obj.GetAnnotations...`? This call will need to be treated as a tuple, either way, yeah?

Reply:
Since we want to test whether the key exists in the map or not, we actually need to do a 2 value assignment (ie. the comma ok idiom). Simply returning without assigning will only result in one value being returned, the value of the map key.

Example:
https://go.dev/play/p/x9NarquVj5X

Reply:
Golang isn't a real language, is it? ;)